src/legacy/core_plugins/kibana/public/discover/plugin.ts

```diff
@@ -33,28 +41,108 @@ import {
+  data: DataPublicPluginStart;
+  share: SharePluginStart;
+  inspector: any;
 }
+const innerAngularName = 'app/discover';
+const embeddableAngularName = 'app/discoverEmbeddable';
 
+/**
+ * Contains Discover, one of the oldest parts of Kibana
+ * There are 2 kinds of Angular bootstrapped for rendering, additionally to the main Angular
+ * Discover provides embeddables, those contain a slimmer Angular
+ */
 export class DiscoverPlugin implements Plugin<DiscoverSetup, DiscoverStart> {
+  private servicesInitialized: boolean = false;
+  private innerAngularInitialized: boolean = false;
+  /**
+   * why are those functions public? they are needed for some mocha tests
+   * can be removed once all is Jest
+   */
+  public initializeInnerAngular?: () => void;
+  public initializeServices?: () => void;
   constructor(initializerContext: PluginInitializerContext) {}
   setup(core: CoreSetup, plugins: DiscoverSetupPlugins): DiscoverSetup {
-    registerFeature();
-    require('./angular');
+    plugins.kibana_legacy.registerLegacyApp({
+      id: 'discover',
+      title: 'Discover',
+      order: -1004,
+      euiIconType: 'discoverApp',
+      mount: async (context, params) => {
+        if (!this.initializeServices) {
+          throw Error('Discover plugin method initializeServices is undefined');
+        }
+        if (!this.initializeInnerAngular) {
+          throw Error('Discover plugin method initializeInnerAngular is undefined');
+        }
+        await this.initializeServices();
+        await this.initializeInnerAngular();
+        const { renderApp } = await import('./application');
+        return renderApp(innerAngularName, params.element);
+      },
+    });
   }
 
   start(core: CoreStart, plugins: DiscoverStartPlugins): DiscoverStart {
-    // TODO enable this when possible, seems it broke a functional test:
-    // dashboard mode Dashboard View Mode Dashboard viewer can paginate on a saved search
-    // const factory = new SearchEmbeddableFactory(plugins.uiActions.executeTriggerActions);
-    // plugins.embeddable.registerEmbeddableFactory(factory.type, factory);
+    // we need to register the application service at setup, but to render it
+    // there are some start dependencies necessary, for this reason
+    // initializeInnerAngular + initializeServices are assigned at start and used
+    // when the application/embeddable is mounted
+    this.initializeInnerAngular = async () => {
+      if (this.innerAngularInitialized) {
+        return;
+      }
+      // this is used by application mount and tests
+      const module = getInnerAngularModule(innerAngularName, core, plugins);
+      setAngularModule(module);
+      this.innerAngularInitialized = true;
+    };
+
+    this.initializeServices = async (test = false) => {
+      if (this.servicesInitialized) {
+        return;
+      }
+      const services = await buildServices(core, plugins, test);
+      setServices(services);
+      this.servicesInitialized = true;
+    };
```

Comment:
I believe this could all be moved into the `mount` function once https://github.com/elastic/kibana/pull/50231 is merged. But this makes sense for now. Another way you could do this is by just keeping a reference to `CoreStart` and `DiscoverStartPlugins` and then passing them into `renderApp`. The advantage here would be that you could defer loading angular to the point where the app is about to be mounted.

Reply:
`initializeServices` is needed  for application `mount` and for embeddable rendering. Furthermore `initializeInnerAngular` is needed for `mocha` tests, so we can definitely improve that once all is Jest. However. since these functions are only executed when the application is loaded / the embeddable needs an Angular injector, loading angular is deferred. When the embeddable uses our data grid in the future, we can clean up here